### PR TITLE
Remove all global attributes from TextureAtlas implementation

### PR DESCRIPTION
**Pull request by SimonDanisch**
That our texture atlas relied on globals for setting e.g. the resolution/glyph_padding/glyph_per_pixel, was always kind of insane, considering we're allowing backends to use different texture atlases (e.g. for smaller sizes in WGLMakie).
I suspected that this also causes https://github.com/MakieOrg/Makie.jl/issues/2480, but that turned out to be false so far :( although `.` does work now, but the marker is still rendered incorrectly! So looks like something goes wrong in FreeTypeAbstraction in the new Julia version or with caching binary.

I've also started using our own serialization format which may remove some invalidations making it easier to compile getting the texture atlas (lets see what the benchmarks say). But what it definitely helps with is, that we can now download the serialized format for all Julia versions and never need to cache fonts again :) I hope this will make a positive dent in the CI runtimes. 

Todo:

- [x] use a stable hash

### CODE REVIEW

**Comment by SimonDanisch** [comment]
Ok, I finally fixed the issue with https://github.com/JuliaLang/julia/pull/47184#pullrequestreview-1163539519 by, *drumroll*, changing `Float16` in https://github.com/MakieOrg/Makie.jl/blob/master/src/utilities/texture_atlas.jl#L284, to `Float32`...
I can confirm the observation, that also just re-evaling the function fixes the issue as well...
I'm guessing we have some problems emitting function signatures for `Float16` in the image?
I put the code_native in diff checker, but only found the function counters to change between the working and buggy version:
https://www.diffchecker.com/ou3VbpAO
@vchuravy do you have any ideas? :sweat_smile: 
MWE: 
```julia
using GLMakie
atlas = Makie.get_texture_atlas();
p = Makie.render_path(Makie.to_spritemarker(:utriangle));
heatmap(Makie.sdistancefield(p, 5, 12), colorrange=(-40, 40)) # will be a flat image when not working

open("not-faulty.txt", "w") do io
    code_native(io, Makie.sdistancefield, Tuple{Matrix{UInt8}, Int, Int}, debuginfo=:none)
end
```
Should also Work with older makie versions, I suppose!

**Comment by SimonDanisch** [comment]
Anyone against changing the arrow marker in `arrows` from `'▲'` to `:utriangle`? This needs a reference image update...

**Comment by SimonDanisch** [comment]
This seems to be a real change:
[Screencast from 2022-12-15 19-50-10.webm](https://user-images.githubusercontent.com/1010467/207943118-65a1bfd4-4a36-4e44-9f8b-2454e8e27052.webm)
@ffreyer how could we best check if this is a regression? 
I wouldn't be surprised, if I actually fixed some bug with all these globals being referenced before..
Of course there's also a good chance I messed something up... We don't have tests checking the alignment in a way that's immediately visible, do we?

**Comment by ffreyer** [comment]
I remember counting pixels for a few sizes, probably with rect marker. I guess we could check the exact sub-matrix for a marker in the atlas (width, height, values at a given pixelsize and font), and the relevant stuff GLMakie etc produces (uvs, quad scale, position)?

File: src/utilities/texture_atlas.jl [hunk 1]
```diff
@@ -1,150 +1,253 @@
...
+function TextureAtlas(; resolution=2048, pix_per_glyph=64, glyph_padding=12, downsample=5)
     return TextureAtlas(
-        RectanglePacker(Rect2(0, 0, initial_size...)),
-        Dict{Tuple{UInt64, String}, Int}(),
-        1,
+        RectanglePacker(Rect2{Int32}(0, 0, resolution, resolution)),
+        Dict{UInt32, Int}(),
         # We use float max here to avoid texture bleed. See #2096
-        fill(Float16(0.5PIXELSIZE_IN_ATLAS[] + GLYPH_PADDING[]), initial_size...),
+        fill(Float16(0.5pix_per_glyph + glyph_padding), resolution, resolution),
```
Comment (ffreyer) [inline src/utilities/texture_atlas.jl:56]: Seems like I didn't update this when I switched away from floatmax. 

The fill value there is the maximum distance from the center of a glyph. With stroke- and glowwidth the background can become relevant during rendering and floatmax to some value can become a harsh cut-off. With this intermediate value it's less harsh (though still there). Basically this is closer to the values you'd get with more padding, which gives you softer transitions with large glow/strokewidths.

File: src/utilities/texture_atlas.jl [hunk 2]
```diff
@@ -396,86 +439,86 @@ primitive_offset(x, scale) = scale ./ -2f0  # default offset
...
 # regardless of its true size.
-function marker_scale_factor(char::Char, font)
-    ta = Makie.get_texture_atlas()
-    lbrt = glyph_uv_width!(ta, char, font)
+function marker_scale_factor(atlas::TextureAtlas, char::Char, font)
+    lbrt = glyph_uv_width!(atlas, char, font)
     uv_width = Vec(lbrt[3] - lbrt[1], lbrt[4] - lbrt[2])
-    full_pixel_size_in_atlas = uv_width .* Vec2f(size(ta.data) .- 1)
-    return full_pixel_size_in_atlas ./ Makie.PIXELSIZE_IN_ATLAS[]
+    full_pixel_size_in_atlas = uv_width .* Vec2f(size(atlas) .- 1)
```
Comment (ffreyer) [inline src/utilities/texture_atlas.jl:459]: This should be `size(atlas)` now as well, I think. Same with other divisions or multiplications by `size(atlas) - 1)`.

File: src/utilities/texture_atlas.jl [hunk 3]
```diff
@@ -180,81 +283,35 @@ function glyph_uv_width!(atlas::TextureAtlas, glyph, font::NativeFont)
...
+        uv_pixel = render(atlas, path_or_glyp)
+        tex_size = Vec2f(size(atlas))
         # 0 based
         idx_left_bottom = minimum(uv_pixel)
         idx_right_top = maximum(uv_pixel)
-
         # transform to normalized texture coordinates
         # -1 for indexing offset
         uv_left_bottom_pad = (idx_left_bottom) ./ tex_size
         uv_right_top_pad = (idx_right_top .- 1) ./ tex_size
```
[the remark this answers is not in the document]
Comment (SimonDanisch) [reply src/utilities/texture_atlas.jl:311]: Could the `#define ANTIALIAS_RADIUS 0.8` be relevant in that context?

[the remark this answers is not in the document]
Comment (ffreyer) [reply src/utilities/texture_atlas.jl:311]: 0.8 works better though it doesn't seem to be sensitive to what ANTIALIAS_RADIUS is set to...

[the remark this answers is not in the document]
Comment (ffreyer) [reply src/utilities/texture_atlas.jl:311]: I currently have it set to 0 in sprites.geom and distance_shape.frag, testing with `p = BezierPath([MoveTo(0, 0), LineTo(0.5, 0.5), LineTo(0.5, -0.5), LineTo(-0.5, -0.5)])` to throw in something else.

[the remark this answers is not in the document]
Comment (ffreyer) [reply src/utilities/texture_atlas.jl]: Is 0.8 also better in WGLMakie? I only checked for GLMakie

**Comment by vchuravy** [comment]
>  @vchuravy do you have any ideas? 😅

If you can narrow it down to the behavior of a particular function that would help.

code_llvm will sadly lie to you and not give you the version that went into the pkgimage

**Comment by ffreyer** [comment]
To explain the changes I made:

The `uv_right_top_pad = (idx_right_top .- 0.8) ./ tex_size` is now just `uv_right_top_pad = idx_right_top ./ tex_size`. With this the uv coordinates are now flush with the outer pixel borders. Before we were excluding one pixel. For reference, this shows the texture atlas in a 0..1 range like opengl uses it with the uv rect of the "L" glyph. Before this was cutting out 80% (100% with -1) of the right/bottom part of the glyph.

```julia
ta = Makie.get_texture_atlas()
fig, ax, p = image(0..1, 0..1, ta.data, interpolate = false)
uv = Makie.glyph_uv_width!(ta, 'L', Makie.defaultfont())
x0, y0, x1, y1 = uv
r = Rect2f(x0, y0, x1-x0, y1-y0)
lines!(ax, r, color = :red)
fig
```

image

I also `get_uv_img` to match these changes.

The second problem here is with the downsampling in `sdistancefield` . If the input image is not dividable by `downsample` we expand (ceil) the image, which effectively shrinks the data shown in it. Using floor instead would shrink the image, making the magnifying the data.

For Bezierpaths we can avoid this problem by matching the generated image size to `downsample` which I did. This should fix the issues with Bezierpath markers being a couple pixels too small (or to large with floor) at large markersizes.

I'm not sure if this is a problem for characters but I guess it probably is. It might be more correct to set `pix_per_glyph` to a multiple of 5, or maybe it doesn't matter because the rendered character is an arbitrary size image?

**Comment by ffreyer** [comment]
Simon noticed that glow didn't work correctly anymore:
Screenshot from 2022-12-20 15-45-10

Afaik when opengl samples near the edge of a pixel it'll interpolate between that pixel and the neighboring pixel. So when we set the uv's for a glyph tight to the outer edges in the texture atlas, it'll consider the pixels around the glyph as well. (I forgot about this yesterday...) 
Now in this case the glyph had other glyphs left and right of it, but not above and below. So the glow value decayed faster up and down. To fix that I moved the border to pixel centers, which does seem to fix that.

For reference, the uv Rect of the scatter marker now looks like this:
Screenshot from 2022-12-20 15-37-14

I guess this should also fix the line artifacts that are sometimes visible with glow and make the background value entirely irrelevant as we never interpolate to it!?

**Comment by SimonDanisch** [comment]
Ok, I think this is finally ready! Thanks a ton @ffreyer! WGLMakie looks a bit broken, but I think that has been the case before and should get fixed in #2428.
@vchuravy, sorry, I'm running out of time here to reproduce this more succinctly :( I did try for a while, but it's really flaky.
E.g., `Makie.TextureAtlas().data` was producing the wrong values as well, which seemed like a great target for an MWE, since it's just:
```julia
function TextureAtlas(; resolution=2048, pix_per_glyph=64, glyph_padding=12, downsample=5)
    return TextureAtlas(
        RectanglePacker(Rect2{Int32}(0, 0, resolution, resolution)),
        Dict{UInt32, Int}(),
        # We use the maximum distance of a glyph as a background to reduce texture bleed. See #2096
        fill(Float16(0.5pix_per_glyph + glyph_padding), resolution, resolution),
        Vec4f[],
        pix_per_glyph,
        glyph_padding,
        downsample,
        Function[]
    )
end
```
and should return data == `fill(Float16(44), resolution, resolution)` but from time to time it returned `fill(Float16(2.29e-5), resolution, resolution)`... It always filled it with the same incorrect value, but as soon as I started trying to make an MWE or change something unrelated and recompile, it would start returning 44 correctly :shrug:
Not sure what other kind of global state could be involved here to create this flaky behavior.

**Pull request merged**